### PR TITLE
GH-42222: [Python] Add bindings for CopyTo on RecordBatch and Array classes

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
### Rationale for this change

We have added bindings for the Device and MemoryManager classes (https://github.com/apache/arrow/issues/41126), and as a next step we can expose the functionality to copy a full Array or RecordBatch to a specific memory manager.

### What changes are included in this PR?

This adds a `copy_to` method on pyarrow Array and RecordBatch.

### Are these changes tested?

Yes

* GitHub Issue: #42222

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
I am not entirely sold on the API on the python side. While this currently 1:1 exposes the C++ `CopyTo` method as a `copy_to()` method, on the Python side we could also have just a `copy()` method (and for example the default memory manager could be the one of the calling object). However, ideally such a method could (optionally?) truncate sliced buffers and child arrays (xref https://github.com/apache/arrow/issues/37878, https://github.com/apache/arrow/issues/388060).

That requested functionality does not yet exist, though. But assuming we add this in the near future, I am not sure it would be great to have both a `copy` and `copy_to` method, as it feels those could be integrated into a single method on the pytho side.

**Comment by danepitkin** [comment]
Overall, the current changes LGTM!

I do like your proposal of having one API. Would your alternative option be an API in the MemoryManager interface e.g. `new_arr = MemoryManager.copy(array)`?

File: python/pyarrow/tests/test_device.py [hunk 1]
```diff
@@ -41,3 +41,16 @@ def test_buffer_device():
...
     assert buf.device == pa.default_cpu_memory_manager().device
     assert buf.memory_manager.is_cpu
+
+
+def test_copy_to():
+    mm = pa.default_cpu_memory_manager()
+    arr = pa.array([0, 1, 2])
+    arr_copied = arr.copy_to(mm)
+    assert arr.equals(arr_copied)
+    assert arr.buffers()[1].address != arr_copied.buffers()[1].address
```
Comment (pitrou) [inline python/pyarrow/tests/test_device.py:51]: Can we also check that the device and memory manager are as expected?

File: python/pyarrow/table.pxi [hunk 2]
```diff
@@ -3549,6 +3549,29 @@ cdef class RecordBatch(_Tabular):
                                                                              row_major, pool))
         return pyarrow_wrap_tensor(c_tensor)
 
+    def copy_to(self, MemoryManager memory_manager):
```
Comment (pitrou) [inline python/pyarrow/table.pxi:3552]: We could perhaps also accept a Device and copy to the Device's default memory manager, for usability.

File: python/pyarrow/table.pxi [hunk 3]
```diff
@@ -3549,6 +3549,29 @@ cdef class RecordBatch(_Tabular):
...
+
+        Returns
+        ------
+        RecordBatch
+        """
+        cdef:
+            shared_ptr[CRecordBatch] c_batch
+
+        with nogil:
+            c_batch = GetResultValue(self.batch.CopyTo(memory_manager.unwrap()))
```
Comment (pitrou) [inline python/pyarrow/table.pxi:3572]: Can `memory_manager` be None here?

**Comment by jorisvandenbossche** [comment]
> I am not entirely sold on the API on the python side. While this currently 1:1 exposes the C++ `CopyTo` method as a `copy_to()` method, on the Python side we could also have just a `copy()` method

In addition, on the C++ side there is also `ViewOrCopyTo` that we should expose in Python, and so the question becomes if that should be a separate method or if that could also be combined in a single API.

File: python/pyarrow/table.pxi (hunk 3 above)
Comment (jorisvandenbossche) [reply python/pyarrow/table.pxi:3592]: Ah, yes, with the current design it is a required keyword and so I should protect it from not being None (add a `not None` in the signature). 

But we might want to allow it to be None, and in that case use the source's memory manager as destination device (i.e. copy to the same device)?

**Comment by jorisvandenbossche** [comment]
@pitrou do you have any thoughts on the API design here?

- Either we simply mimic the C++ API exactly, i.e. add a `copy_to()` and `view_or_copy_to()` on Array/RecordBatch (although then on the Buffer class it is `copy()` and not `copy_to()` if we follow C++ .., which is a bit inconsistent)
- or add a bit more typical python `copy()` method with an optional device argument (and default would then be to copy to the same device). But question is then if we would include the functionality of `ViewOrCopyTo` in that same method? (at which point the result could be only a "shallow" copy, i.e. same buffers in new Array/RecordBatch object) In any case we can't add a `view()` method as that is already taken on Array (for viewing as another data type)

**Comment by jorisvandenbossche** [comment]
As another idea based on the Array API (although then deviating much more from the C++ API) could be a `to_device()` method (https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.to_device.html), which could quite naturally combine the functionality of `CopyTo` and `ViewOrCopyTo`, I think.

File: python/pyarrow/table.pxi (hunk 3 above)
Comment (pitrou) [reply python/pyarrow/table.pxi:3592]: I don't think copying to the same memory manager is really useful, so we needn't allow None IMHO.

**Comment by pitrou** [comment]
My preference would be on adding `copy_to()` and `view_or_copy_to()` methods.

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit test-cuda-python

**Comment by jorisvandenbossche** [comment]
> My preference would be on adding `copy_to()` and `view_or_copy_to()` methods.

OK, updated the PR and keeping the `copy_to` name then (and will add `view_or_copy_to()`) as well).

I don't have a strong opinion about the names so happy to go with those. I think my main concern is that IMO it is likely we are going to want to add a `Array.copy()` method at some point as well (that copies with truncation of sliced buffers), and then we have three similar method names.

**Comment by danepitkin** [comment]
LGTM! There are just some numpydoc errors left

**Comment by jorisvandenbossche** [comment]
@gitbub-actions crossbow submit test-conda-python-3.12

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit test-conda-python-3.12

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit wheel-windows-cp312-amd64

File: python/pyarrow/device.pxi [hunk 4]
```diff
@@ -64,6 +64,9 @@ cdef class Device(_Weakrefable):
         self.init(device)
         return self
 
+    cdef inline shared_ptr[CDevice] unwrap(self) nogil:
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply python/pyarrow/device.pxi:67]: In practice it's what we do for almost all our `unwrap` functions (and the few that don't use it probably should for consistency)

But the reason is 1) that this function does not need to GIL (there is no Python interaction), so it _can_ be marked with `nogil`, and 2) actually marking it as such ensures that we can call this method from within a `with nogil: ...` block (and given that calling `unwrap()` is typically needed when calling some C++ function, and we typically use `with nogil: ...` blocks when we are calling a C++ function, in practice there are quite some cases where we indeed use `unwrap()` inside such a block)

[the remark this answers is not in the document]
Comment (danepitkin) [reply python/pyarrow/device.pxi:67]: The first section in this doc really helped me understand when/why to release the gil: https://cython.readthedocs.io/en/latest/src/userguide/nogil.html

**Comment by jorisvandenbossche** [comment]
I will leave `copy_or_view_to` for a follow-up issue/PR, so I think this one is ready now

**Review by danepitkin — APPROVED** [review]

**Pull request merged**